ApiToken docs and filter update. (#417)
codeigniter4/shield: docs/guides/api-tokens.md

# Protecting an API with Access Tokens

Access Tokens can be used to authenticate users for your own site, or when allowing third-party developers to access your API. When making requests using access tokens, the token should be included in the `Authorization` header as a `Bearer` token.

Tokens are issued with the `generateAccessToken()` method on the user. This returns a `CodeIgniter\Shield\Entities\AccessToken` instance. Tokens are hashed using a SHA-256 algorithm before being saved to the database. The access token returned when you generate it will include a `raw_token` field that contains the plain-text, un-hashed, token. You should display this to your user at once so they have a chance to copy it somewhere safe, as this is the only time this will be available. After this request, there is no way to get the raw token.

> The `generateAccessToken` method requires a name for the token. These are free strings and are often used to identify the user/device the token was generated from, like 'Johns MacBook Air'.

Reviewer: ```suggestion
The `generateAccessToken()` method requires a name for the token. These are free strings and are often used to identify the user/device the token was generated from, like 'Johns MacBook Air'.
```

```php
$routes->get('/access/token', static function() {

>     $token = auth()->user()->generateAccessToken(request()->getVar('token_name));

Reviewer: `request()` ?
Do you mean `service('request')` ?

Reviewer 2: Don't mind him, [Laravel](https://laravel.com/docs/9.x/helpers#method-request) gets the bills paid. Yes he meant to use the service.

Author: Dang it :) I've considered adding a `request()` and `response()` helper method for years on this framework, but never did. I actually had to go double-check the code to ensure we didn't have it lol. But, yes, that's what I meant.

    return json_encode(['token' => $token->raw_token]);
});
```

You can access all of the user's tokens with the `accessTokens()` method on that user.

```php
$tokens = $user->accessTokens();
foreach($tokens as $token) {
    //
}
```

## Token Permissions

Access tokens can be given `scopes`, which are basically permission strings, for the token. This is generally not the same as the permission the user has, but is used to specify the permissions on the API itself. If not specified, the token is granted all access to all scopes. This might be enough for a smaller API.

```php
return $user->generateAccessToken('token-name', ['users-read'])->raw_token;
```

> NOTE: At this time, scope names should avoid using a colon (:) as this causes issues with the route filters being correctly recognized.

Reviewer: ```suggestion
> **Note**
> At this time, scope names should avoid using a colon (`:`) as this causes issues with the route filters being correctly recognized.
```

> When handling incoming requests you can check if the token has been granted access to the scope with the `tokenCan` method.

Reviewer: ```suggestion
When handling incoming requests you can check if the token has been granted access to the scope with the `tokenCan()` method.
```

```php
if ($user->tokenCan('users-read')) {
    //
}
```

### Revoking Tokens

Tokens can be revoked by deleting them from the database with the `revokeAccessToken($rawToken)` or `revokeAllAccessTokens()` methods.

```php
$user->revokeAccessToken($rawToken);
$user->revokeAllAccessTokens();
```

## Protecting Routes

The first way to specify which routes are protected is to use the `tokens` controller filter.

For example, to ensure it protects all routes under the `/api` route group, you would use the `$filters` setting on `app/Config/Filters.php`.

```php
public $filters = [
    'tokens' => ['before' => ['api/*']],
];
```

You can also specify the filter should run on one or more routes within the routes file itself:

```php
$routes->group('api', ['filter' => 'tokens'], function($routes) {
    //
});
$routes->get('users', 'UserController::list', ['filter' => 'tokens:users-read']);
```

When the filter runs, it checks the `Authorization` header for a `Bearer` value that has the raw token. It then hashes the raw token and looks it up in the database. Once found, it can determine the correct user, which will then be available through an `auth()->user()` call.

> Note: Currently only a single scope can be used on a route filter. If multiple scopes are passed in, only the first one is checked.

Reviewer: ```suggestion
> **Note**
> Currently only a single scope can be used on a route filter. If multiple scopes are passed in, only the first one is checked.
```

# Protecting an API with Access Tokens

Access Tokens can be used to authenticate users for your own site, or when allowing third-party developers to access your API. When making requests using access tokens, the token should be included in the `Authorization` header as a `Bearer` token.

> To issue tokens for users, the `UserModel` must use the `CodeIgniter\Shield\Authentication\Traits\HasAccessTokens` trait. The `UserModel` that ships with Shield already uses this trait.
> 
> ```php
> use CodeIgniter\Shield\Authentication\Traits\HasAccessTokens
> use CodeIgniter\Model;
> 
> class UserModel extneds Model
> {
>     use HasAccessTokens;
> }
> ```
> 

Reviewer 2: We now require all User models to be an extension of ours, so this will always be the case. This can be dropped:
```suggestion
```

Tokens are issued with the `generateAccessToken()` method on the user. This returns a `CodeIgniter\Shield\Entities\AccessToken` instance. Tokens are hashed using a SHA-256 algorithm before being saved to the database. The access token returned when you generate it will include a `raw_token` field that contains the plain-text, un-hashed, token. You should display this to your user at once so they have a chance to copy it somewhere safe, as this is the only time this will be available. After this request, there is no way to get the raw token.

The `generateAccessToken` method requires a name for the token. These are free strings and are often used to identify the user/device the token was generated from, like 'Johns MacBook Air'.

```php
$routes->get('/access/token', static function() {
    $token = auth()->user()->generateAccessToken(request()->getVar('token_name));

>     return ['token' => $token->raw_token];

Reviewer 2: CodeIgniter doesn't have fancy array response handling 😉 You've been hanging around "those other guys" too much. 
```suggestion
    return json_encode(['token' => $token->raw_token]);
```

Author: Lol I know. Darn day jobs :)

});
```

> You can access all of the users' tokens with the `accessTokens()` method on the user.

Reviewer 2: ```suggestion
You can access all of the user's tokens with the `accessTokens()` method on that user.
```

```

NOTE: At this time, scope names should avoid using a colon (:) as this causes issues with the route filters being correctly recognized.

When handling incoming requests you can check if the token has been granted access to the scope with the `tokenCan` method.

```php
if ($user->tokenCan('users-read')) {
    //
}
```

### Revoking Tokens

Tokens can be revoked by deleting them from the database with the `revokeAccessToken($rawToken)` or `revokeAllAccessTokens()` methods.

```php
$user->revokeAccessToken($rawToken);
$user->revokeAllAccessTokens();
```

## Protecting Routes

The first way to specify which routes are protected is to use the `tokens` controller filter.

For example, to ensure it protects all routes under the `/api` route group, you would use the `$filters` setting on `app/Config/Filters.php`.

```php
public $filters = [
    'tokens' => ['before' => ['api/*']],
];
```

You can also specify the filter should run on one or more routes within the routes file itself:

```php
$routes->group('api', ['filter' => 'tokens'], function($routes) {
    //
});
$routes->get('users', 'UserController::list', ['filter' => 'tokens:users-read']);
```

> When the filter runs, it checks the `Authorization` header for a `Bearer` value that has the raw token. It then looks hashes the raw token and looks it up in the database. Once found, it can determine the correct user, which will then be available through an `auth()->user()` call.

Reviewer 2: ```suggestion
When the filter runs, it checks the `Authorization` header for a `Bearer` value that has the raw token. It then hashes the raw token and looks it up in the database. Once found, it can determine the correct user, which will then be available through an `auth()->user()` call.
```
